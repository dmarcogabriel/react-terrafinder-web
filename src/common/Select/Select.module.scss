@import 'styles/';

.selectInput {
  position: relative;
  min-width: 10rem;

  .button {
    padding: $size-sm $size-sm;
    background: $white;
    border-radius: $size-md;
    width: 100%;
    cursor: pointer;

    .label {
      font-size: $size-sm;
      color: $orange;
    }

    .content {
      display: flex;
      flex-direction: row;
    }
  }
}

.selectOptionList {
  position: absolute;
  background: $white;

  .selectOption {
    width: 10rem;

    .selectOptionButton {
      display: flex;
      justify-content: center;
      width: 100%;
      padding: $size-sm 0;
      font-size: $size-xs;
      background: $white;
      color: $orange;
      cursor: pointer;
    }

    .selected {
      background: $orange;
      color: $white;
    }
  }
}

.selectArrowIcon {
  color: $orange;
}
