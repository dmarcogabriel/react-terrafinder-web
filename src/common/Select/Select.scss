@import 'styles/';

.selectInput {
  position: relative;
  min-width: 10rem;

  .searchInputButton {
    padding: $size-sm $size-sm;
    background: $white;
    width: 100%;
    cursor: pointer;

    .searchInputLabel {
      font-size: $size-sm;
      color: $orange;
    }

    .searchInputContent {
      display: flex;
      flex-direction: row;
    }
  }
}

.selectOptionList {
  min-width: 100%;
  position: absolute;
  background: $white;
  z-index: 50;
  box-shadow: 0 2px 10px 1px rgba(0, 0, 0, 0.5);
  padding: 1rem 0;
  display: flex;
  align-items: center;
  justify-content: center;

  .selectOptionButton {
    font-size: 1rem;
    background: $white;
    color: $orange;
    cursor: pointer;
  }

  .selectedOption {
    background: $orange;
    color: $white;
  }
}

.selectArrowIcon {
  color: $orange;
}
