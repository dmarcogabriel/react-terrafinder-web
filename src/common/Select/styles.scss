@import '../../styles/';

.selectInput {
  position: relative;
  width: 10rem;

  .selectInput--button {
    padding: $size-sm $size-xs;
    background: $white;
    border-radius: $size-md;
    cursor: pointer;

    .selectInput--button--label {
      font-size: $size-xs;
      color: $orange;
    }

    .selectInput--button--content {
      display: flex;
      flex-direction: row;
    }
  }
}

.selectOptionList {
  position: absolute;
  background: $white;

  .selectOptionList--option {
    width: 10rem;

    .selectOptionList--optionButton {
      display: flex;
      justify-content: center;
      width: 100%;
      padding: $size-sm 0;
      font-size: $size-xs;
      background: $white;
      color: $orange;
      cursor: pointer;
    }

    .selectOptionList--optionButton-selected {
      background: $orange;
      color: $white;
    }
  }
}
