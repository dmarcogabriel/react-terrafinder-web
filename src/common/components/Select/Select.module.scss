@import 'styles';

.selectInput {
  position: relative;

  > button {
    padding: $size-sm $size-sm;
    background: $white;
    width: 100%;
    cursor: pointer;

    .searchInputLabel {
      color: $gray-light;
      margin: 0.5rem 0;
    }

    .searchInputContent {
      display: flex;
      flex-direction: row;
      justify-content: space-between;
      align-items: center;
      border: 1px solid $gray-light;
      border-radius: 0.25rem;
      padding: 0.8rem 0.5rem;

      .selectArrowIcon {
        color: $gray-light;
      }
    }
  }

  .selectOptionList {
    position: absolute;
    bottom: -2.5rem;
    left: 1rem;
    right: 1rem;
    z-index: 50;
    box-shadow: $shadow-base;
    border-bottom-left-radius: 0.25rem;
    border-bottom-right-radius: 0.25rem;
    overflow: hidden;

    > li {
      > button {
        display: flex;
        align-items: center;
        justify-content: center;
        padding: 1rem 0;
        cursor: pointer;
        color: $gray-dark;
        z-index: 51;
        background: $white;
        width: 100%;
      }

      .selected {
        background: $orange;
        width: 100%;
      }
    }
  }
}
