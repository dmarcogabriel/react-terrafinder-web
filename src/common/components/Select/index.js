import React, { useState, useEffect } from 'react';
import { MdArrowDropUp, MdArrowDropDown } from 'react-icons/md';
import classes from './Select.module.scss';

export default function Select({
  label,
  value,
  options = [],
  id,
  onChange = () => {},
}) {
  const [showOptions, setShowOptions] = useState(false);
  const [defaultValue, setDefaultValue] = useState(value || 'Selecione...');
  const [optionsList, setOptionsList] = useState(options);

  const handleClickOutside = () => {
    // todo: handle click outside
    // const { classList } = e.target;
    // console.log(classList.contains('Select_option'));
    // if (!classList.contains('Select_option')) {
    //   setShowOptions(false);
    // }
  };

  const handleSelectOption = (e) => {
    setOptionsList(
      optionsList.map((option) => ({
        ...option,
        selected: option.key === e.key,
      }))
    );
    setDefaultValue(e.name);
    setShowOptions(false);
    onChange(e);
  };

  useEffect(() => {
    // document.addEventListener('mousedown', handleClickOutside);
    // return () => document.removeEventListener('mousedown', handleClickOutside);
  });

  return (
    <div className={classes.selectInput}>
      <button type="button" onClick={() => setShowOptions(!showOptions)}>
        <p className={classes.searchInputLabel} label={id}>
          {label}
        </p>

        <div className={classes.searchInputContent}>
          <p>{defaultValue}</p>

          {showOptions ? (
            <MdArrowDropUp className={classes.selectArrowIcon} size={22} />
          ) : (
            <MdArrowDropDown className={classes.selectArrowIcon} size={22} />
          )}
        </div>
      </button>

      {showOptions && (
        <ul className={classes.selectOptionList} id={id}>
          {optionsList.map((option) => (
            <li key={option.key}>
              <button
                type="button"
                className={option.selected && classes.selected}
                onClick={() => handleSelectOption(option)}
              >
                {option.name}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
}
