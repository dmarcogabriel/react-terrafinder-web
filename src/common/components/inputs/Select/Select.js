import React from 'react';
import {
  Box,
  Select,
  InputLabel,
  MenuItem,
  FormHelperText,
} from '@mui/material';
import {
  Mood as ValidIcon,
  SentimentDissatisfied as InvalidIcon,
} from '@mui/icons-material';

const ValidationIcon = ({ errorMessage }) =>
  errorMessage ? (
    <InvalidIcon fontSize="small" color="error" />
  ) : (
    <ValidIcon fontSize="small" color="success" />
  );

export const SelectInput = ({
  value,
  label,
  onChange,
  dataTestId = 'select-input',
  options,
  variant = 'standard',
  noValidation,
  errorMessage,
  isLoading,
  inputProps = {},
  ...props
}) => {
  const handleChange = ({ target: { value: selectedOption } }) => {
    onChange(selectedOption);
  };

  return (
    <Box
      variant={variant}
      sx={{
        width: '100%',
        my: 2,
      }}
    >
      <Box
        sx={{
          display: 'flex',
          alingItems: 'center',
          justifyContent: 'space-between',
        }}
      >
        <InputLabel error={!!errorMessage}>{label}</InputLabel>
        {!noValidation && <ValidationIcon errorMessage={errorMessage} />}
      </Box>
      <Select
        inputProps={{ 'data-testid': dataTestId, ...inputProps }}
        value={isLoading ? 'Carregando...' : value}
        onChange={handleChange}
        fullWidth
        size="small"
        error={!!errorMessage}
        disabled={isLoading}
        {...props}
      >
        {!isLoading &&
          options.map((option, i) => (
            <MenuItem
              data-testid={`option-${i}`}
              key={String(i)}
              value={option}
            >
              {option}
            </MenuItem>
          ))}
      </Select>
      {!!errorMessage && (
        <FormHelperText sx={{ ml: '14px' }} error>
          {errorMessage}
        </FormHelperText>
      )}
    </Box>
  );
};
