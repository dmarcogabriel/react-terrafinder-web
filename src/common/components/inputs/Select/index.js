import { SelectInput } from './Select';

export { SelectInput };
