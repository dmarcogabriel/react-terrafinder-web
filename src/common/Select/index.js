import React, { useState, useEffect } from 'react';
import { MdArrowDropUp, MdArrowDropDown } from 'react-icons/md';
import { useTheme } from 'styled-components';
import cn from 'classnames';
import './styles.scss';

export default function Select({
  label,
  value,
  options = [],
  id,
  onChange = () => {},
}) {
  const [showOptions, setShowOptions] = useState(false);
  const [defaultValue, setDefaultValue] = useState(value || 'Selecione...');
  const [optionsList, setOptionsList] = useState(options);
  const { colors } = useTheme();

  const handleClickOutside = (e) => {
    const { classList } = e.target;

    if (
      !classList.contains('option') &&
      !classList.contains('selected-option')
    ) {
      setShowOptions(false);
    }
  };

  const handleSelectOption = (e) => {
    setOptionsList(
      optionsList.map((option) => {
        return option.key === e.key
          ? { ...option, selected: true }
          : { ...option, selected: false };
      })
    );
    setDefaultValue(e.name);
    setShowOptions(false);
    onChange(e);
  };

  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside);

    return () => document.removeEventListener('mousedown', handleClickOutside);
  });

  return (
    <div className="container">
      <button
        className="button"
        type="button"
        onClick={() => setShowOptions(!showOptions)}
      >
        <p label={id}>{label}</p>

        <div className="content">
          <p>{defaultValue}</p>

          {showOptions ? (
            <MdArrowDropUp style={{ color: colors.orange }} />
          ) : (
            <MdArrowDropDown style={{ color: colors.orange }} />
          )}
        </div>
      </button>

      {showOptions && (
        <ul className="optionsList" id={id}>
          {optionsList.map((option) => (
            <li>
              <button
                type="button"
                key={option.key}
                className={cn('option-button', option.selected && 'selected')}
                onClick={() => handleSelectOption(option)}
              >
                {option.name}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
}
