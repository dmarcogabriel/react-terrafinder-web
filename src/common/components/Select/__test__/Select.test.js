import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import Select from '..';

const mockOptions = [
  {
    key: 'o1',
    selected: false,
    name: 'option 1',
  },
  {
    key: 'o2',
    selected: true,
    name: 'option 2',
  },
];

describe('<Select />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Select />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should show options', () => {
    const { getByTestId } = render(<Select options={mockOptions} />);

    const select = getByTestId('select');
    fireEvent.click(select);

    const arrowDropUp = getByTestId('arrowDropUp');
    const options = getByTestId('options');

    expect(arrowDropUp).toBeInTheDocument();
    expect(options).toBeInTheDocument();
  });

  it('should fire change', () => {
    const { getByTestId } = render(
      <Select options={mockOptions} onChange={() => {}} />
    );

    const select = getByTestId('select');
    fireEvent.click(select);

    const option = getByTestId('option-o1');
    fireEvent.click(option);

    const value = getByTestId('value');
    expect(value).toHaveTextContent('option 1');
  });

  it('matches snapshot', () => {
    const tree = renderer.create(<Select />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
