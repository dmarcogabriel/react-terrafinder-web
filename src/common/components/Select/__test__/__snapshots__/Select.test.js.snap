// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> matches snapshot 1`] = `
<div>
  <button
    data-testid="select"
    onClick={[Function]}
    type="button"
  >
    <p />
    <div>
      <p
        data-testid="value"
      >
        Selecione...
      </p>
      <svg
        data-testid="arrowDropDown"
        fill="currentColor"
        height={22}
        size={22}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width={22}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </button>
</div>
`;
